File: iosApp/iosApp/Modules/Alert/AlertPresenter.swift
// Created by web3dgn on 02/05/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import UniformTypeIdentifiers

enum AlertPresenterEvent {
    case dismiss
}

protocol AlertPresenter: AnyObject {
    func present()
    func handle(_ event: AlertPresenterEvent)
}

final class DefaultAlertPresenter {
    private let context: AlertContext
    private weak var view: AlertView!
    private let wireframe: AlertWireframe

    init(
        context: AlertContext,
        view: AlertView,
        wireframe: AlertWireframe
    ) {
        self.context = context
        self.view = view
        self.wireframe = wireframe
    }
}

extension DefaultAlertPresenter: AlertPresenter {

    func present() {
        let viewModel = AlertViewModel(context: context)
        view.update(with: viewModel)
    }

    func handle(_ event: AlertPresenterEvent) {
        switch event {
        case .dismiss:
            wireframe.navigate(to: .dismiss)
        }
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/alert/AlertPresenter.kt
package com.sonsofcrypto.web3walletcore.modules.alert

import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3walletcore.modules.alert.AlertWireframeDestination.Dismiss

sealed class AlertPresenterEvent {
    data class SelectAction(val idx: Int): AlertPresenterEvent()
    object Dismiss: AlertPresenterEvent()
}

interface AlertPresenter {
    fun present()
    fun handle(event: AlertPresenterEvent)
}

class DefaultAlertPresenter(
    private val view: WeakRef<AlertView>,
    private val wireframe: AlertWireframe,
    private val context: AlertWireframeContext
): AlertPresenter {

    override fun present() { updateView() }

    override fun handle(event: AlertPresenterEvent) {
        when (event) {
            is AlertPresenterEvent.SelectAction -> {
                context.onActionTapped?.let { it(event.idx) }
                handle(AlertPresenterEvent.Dismiss)
            }
            is AlertPresenterEvent.Dismiss -> wireframe.navigate(Dismiss)
        }
    }

    private fun updateView() {
        view.get()?.update(AlertViewModel(context))
    }
}
